### PR TITLE
Issue/16

### LINKED ISSUE #16: Create web interface for history
One page with a list of all builds, each build must be clickable to access the more detailed information about that build.

### PR DESCRIPTION
**Pull request by keivanm**
Resolves #16.

Add Class DocumentBuilder to create HTML pages for the web interface. Add tests for DocumentBuilder.
Also change ContinuousIntegrationServer.java to account for these new changes.

### CODE REVIEW

**Review by Kayi1500 — APPROVED** [review]
Looks good. Tested without any issues too :+1:

File: src/main/java/DocumentBuilder.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            "\n" +
+            "</style>";
+    private static String mid = "</head>\n<body>\n";
+    private static String end = "</body>\n</html>";
+    public DocumentBuilder() {
+    }
+
+    /**
+     * Generates an HTML report by using bufferedwriter to write string objects,
+     * formatted to look like HTML syntax, to a file.
```
Comment (grundb) [inline src/main/java/DocumentBuilder.java:68]: Probably outdated, writes to string

File: src/main/java/DocumentBuilder.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            "#previousDiv {\n" +
+            "\n" +
+            "    margin-top: 50px;\n" +
+            "\n" +
+            "}\n" +
+            "\n" +
+            "</style>";
+    private static String mid = "</head>\n<body>\n";
+    private static String end = "</body>\n</html>";
+    public DocumentBuilder() {
```
Comment (grundb) [inline src/main/java/DocumentBuilder.java:63]: Class probably does not need constructor, methods could be static

File: src/main/java/DocumentBuilder.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                            "</div>"
+                    , commit.getCommitID(), commit.getCommitID(), commit.getBuildDate(),
+                    commit.isBuildResult()));
+        }
+        build.append("</div>");
+
+        return start+css+mid+heading+build+end;
+    }
+
+    public String writeBuildDetails(CommitStructure commit){
```
Comment (grundb) [inline src/main/java/DocumentBuilder.java:91]: Missing javadoc

File: src/test/java/DocumentBuilderTest.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+        List<CommitStructure> commits = new ArrayList<>();
+        commits.add(new CommitStructure("23424324432424", "11/02-20",
+                "logloglogsfsssfsfsfsfsffs", true));
+        commits.add(new CommitStructure("79979779977979", "11/02-20",
+                "logloglogsfsssf645646465466sfsffs", true));
+
+        //set system out to an outputstream so it does not print anything when
+        //running tidy.parse(System.in, System.out)
+        System.setOut(new PrintStream(new OutputStream() {
```
Comment (grundb) [inline src/test/java/DocumentBuilderTest.java:35]: Interesting, and I see the point in this, but doesn't it affect subsequent tests as well unless the I/O is reset later? It seems the empty stream could be placed directly in the call to `parse` instead

File: src/test/java/DocumentBuilderTest.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        assertTrue(html.contains("23424324432424"));
+        assertTrue(html.contains("79979779977979"));
+
+        //https://stackoverflow.com/questions/26608365/how-to-validate-html-using-jtidy
+        //http://jtidy.sourceforge.net/apidocs/index.html
+        Tidy tidy = new Tidy();
+        InputStream stream = new ByteArrayInputStream(html.getBytes());
+        tidy.parse(stream, System.out); //parse the html string
+        assertTrue(tidy.getParseErrors() == 0); //the number of errors that occurred in the most recent parse operation.
+        assertTrue(tidy.getParseWarnings() == 0); //the number of warnings that occurred in the most recent parse operation.
```
Comment (grundb) [inline src/test/java/DocumentBuilderTest.java:48]: I guess it is good if warnings are 0, though the javadoc comment for this test says warnings are ignored

File: src/test/java/DocumentBuilderTest.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        String html = db.writeBuildDetails(commit);
+        assertTrue(html.contains("logloglogsfsssfsfsfsfsffs"));
+
+        //https://stackoverflow.com/questions/26608365/how-to-validate-html-using-jtidy
+        //http://jtidy.sourceforge.net/apidocs/index.html
+        Tidy tidy = new Tidy();
+        InputStream stream = new ByteArrayInputStream(html.getBytes());
+        tidy.parse(stream, System.out); //parse the html string
+        assertTrue(tidy.getParseErrors() == 0); //the number of errors that occurred in the most recent parse operation.
+        assertTrue(tidy.getParseWarnings() == 0); //the number of warnings that occurred in the most recent parse operation.
```
Comment (grundb) [inline src/test/java/DocumentBuilderTest.java:77]: Like the previous test the comment states warnings are ignored, I think any solution is fine but should be consistent

File: src/test/java/DocumentBuilderTest.java (hunk 4 above)
Comment (keivanm) [reply src/test/java/DocumentBuilderTest.java:35]: Yes I thought of that too. I'll look into it

File: src/test/java/DocumentBuilderTest.java (hunk 5 above)
Comment (keivanm) [reply src/test/java/DocumentBuilderTest.java:48]: yeah forgot to update the javadoc

File: src/main/java/ContinuousIntegrationServer.java [hunk 7]
```diff
@@ -35,43 +46,55 @@ public void handle(
...
+                response.getWriter().println(html);
+                response.flushBuffer();
 
-            response.setStatus(javax.servlet.http.HttpServletResponse.SC_OK);
-            response.getWriter().println("CI job done");
-        }
+            } else if (target.equals("/commit")) {
+                DocumentBuilder db = new DocumentBuilder();
+                String commitID = request.getParameter("commitID");
+                CommitStructure commit = database.selectSpecificRow(commitID);
```
Comment (kittytinythai) [inline src/main/java/ContinuousIntegrationServer.java:65]: Ensure the returned `CommitStructure` is not `null`, if that's the case, then the response should have the status 404 not found.

**Review by kittytinythai — CHANGES_REQUESTED** [review]
When I run the server and tries to access `http://localhost:8080/builds` via my browser, I get the following error message from the server:

```
java.sql.SQLNonTransientConnectionException: Could not create connection to database server. Attempted reconnect 3 times. Giving up.
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:110)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:89)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:63)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:73)
	at com.mysql.cj.jdbc.ConnectionImpl.connectWithRetries(ConnectionImpl.java:903)
	at com.mysql.cj.jdbc.ConnectionImpl.createNewIO(ConnectionImpl.java:828)
	at com.mysql.cj.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:453)
	at com.mysql.cj.jdbc.ConnectionImpl.getInstance(ConnectionImpl.java:246)
	at com.mysql.cj.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:198)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:677)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:228)
	at MysqlDatabase.connectToDB(MysqlDatabase.java:44)
	at MysqlDatabase.<init>(MysqlDatabase.java:17)
	at ContinuousIntegrationServer.<init>(ContinuousIntegrationServer.java:30)
	at ContinuousIntegrationServer.main(ContinuousIntegrationServer.java:111)
Caused by: com.mysql.cj.exceptions.CJCommunicationsException: Communications link failure
Caused by: com.mysql.cj.exceptions.CJCommunicationsException: Communications link failure


The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:61)
	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:105)
	at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:151)
	at com.mysql.cj.exceptions.ExceptionFactory.createCommunicationsException(ExceptionFactory.java:167)
	at com.mysql.cj.protocol.a.NativeSocketConnection.connect(NativeSocketConnection.java:89)
	at com.mysql.cj.NativeSession.connect(NativeSession.java:144)
	at com.mysql.cj.jdbc.ConnectionImpl.connectWithRetries(ConnectionImpl.java:847)
	... 10 more
Caused by: java.net.ConnectException: Connection refused: connect
	at java.base/sun.nio.ch.Net.connect0(Native Method)
	at java.base/sun.nio.ch.Net.connect(Net.java:493)
	at java.base/sun.nio.ch.Net.connect(Net.java:482)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:588)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:339)
	at java.base/java.net.Socket.connect(Socket.java:603)
	at com.mysql.cj.protocol.StandardSocketFactory.connect(StandardSocketFactory.java:155)
	at com.mysql.cj.protocol.a.NativeSocketConnection.connect(NativeSocketConnection.java:63)
	... 12 more
Caused by: java.net.ConnectException: Connection refused: connect

Connection is null!
Connection is null!
```

**Pull request merged**